### DISCUSSION TITLE
Theming how-to for the new Fancy Menu

### DISCUSSION THREAD

**Discussion by AzumaHazuki** [lxqt/lxqt#2517] [Show and tell]
Hello all! I've been playing with the Fancy Menu plugin from the -git branch today and for the most part really liking it so far. Have also been looking at the code to figure out what themeable elements there are in this thing and how to do it. Here's what I've found, using an example from a WIP theme called Aquativo:

![FancyMenuPlugin](https://github.com/lxqt/lxqt/assets/67122280/ff73bd33-7d73-4ae5-b917-2a86c53fb19f)

And here's the code that produces this (the scrollbars are defined further in the theme file):

```
/* Fancy Menu styling */
#FancyMenu > QWidget { /* need the > here so as not to stomp on scrollbars... */
	color:				#000000;
	background:			#f6f6f6;

	background-image:		url(assets/boxes/bg.png);
	background-repeat:		repeat-xy;

	border:				1px solid #afafaf;
}

#FancyMenu QToolButton {
	color:				#000000;
	background:			transparent;

	border:				2px solid transparent;
	border-image:			url(assets/buttons/toolbutton.png);

	padding:			2px;
	qproperty-iconSize:		24px;
}

#FancyMenu QToolButton:hover {
	border-image:			url(assets/buttons/active.png);
}

#FancyMenu QToolButton:pressed {
	border-image:			url(assets/buttons/hover.png);
}

#FancyMenu QLineEdit {
	color:				#000000;
	background:			transparent;

	border:				1px solid transparent;
	border-image:			url(assets/boxes/entry.png);

	selection-color:		#ffffff;
	selection-background-color:	#1376d3;

	padding:			2px;	
}

#FancyMenu #AppView {
	color:				#000000;
	background:			#ffffff;

	border:				1px solid #afafaf;

	icon-size:			24px;
}

#FancyMenu #AppView::item {
	color:				#000000;
	background:			transparent;

	margin:				1px;
	padding:			3px;
}

#FancyMenu #AppView::item:hover,
#FancyMenu #AppView::item:selected {
	color:				#ffffff;
	background:			#1376d3;
}

#FancyMenu #CategoryView {
	color:				#000000;
	background:			transparent;

	border:				none;

	icon-size:			24px;
}

#FancyMenu #CategoryView::item {
	color:				#000000;
	background:			transparent;

	border:				2px solid transparent;
	border-image:			none;

	padding:			1px;
}

#FancyMenu #CategoryView::item:hover,
#FancyMenu #CategoryView::item:selected {
	border-image:			url(assets/buttons/active.png);
}
```

What I've gleaned from this exercise:
1. There are 2 `QListView` objects in here, which are accessible through the `#CategoryView` and `#AppView` selectors.
2. I can't figure out what the base widget type for the menu itself is, hence that ugly `#FancyMenu > QWidget` hack. I _really_ do not feel comfortable with that so if anyone knows what the actual type is please tell me! Doesn't seem to be `QMenu, QFrame, QWindow, QDialog, QMainWindow`, or even `QAbstractItemView`
3. As of now there's no way to move the `QLineEdit` object. I think it should be in the same horizontal bounding box as the Settings and Power buttons, to be honest, and the top area should optionally show the user's avatar and/or name.
4. It *is* possible to set the icon sizes for both the two list views _and_ the `QToolButton`s. I think this should also be exposed as a setting in the configuration menu.
5. The menus do not track the mouse pointer, unlike basically every other menu ever, which feels really _really_ weird.

**Comment by stefonarch** [comment]
> I can't figure out what the base widget type for the menu itself is, hence that ugly #FancyMenu > QWidget hack. I really do not feel comfortable with that so if anyone knows what the actual type is please

You can use the dom plugin in the panel to retrieve elements. I think you mean `LXQtFancyMenuWindow{`
https://github.com/lxqt/lxqt-themes/blob/c1771105be7d64c0b76585ac607894b8f3c27bbf/themes/ambiance/lxqt-panel.qss#L331
